### PR TITLE
Split storage flattening into a sequence of simpler passes

### PR DESCRIPTION
**Pull request by abadams**
No functional changes intended. This refactoring makes it possible to place passes in between destructuring tuples into scalars and actually flattening multi-dimensional memory to single-dimensional.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -353,6 +353,7 @@ SOURCE_FILES = \
   Tracing.cpp \
   TrimNoOps.cpp \
   Tuple.cpp \
+  TupleSplitting.cpp \
```
Comment (zvookin) [inline Makefile]: Other passes seem to start with the verb. More like "SplitTuples".

**Comment by zvookin** [comment]
Other than considering changing the name "TupleSplitting" to "SplitTuples", LGTM.

**Comment by ronlieb** [comment]
This patch builds fine with g++, however clang++ is detecting a warning which we treat as error
/local/mnt2/ronl/Halide-Masterbuild/Halide/src/StorageFlattening.cpp:28:29: error: private field 'outputs' is
      not used [-Werror,-Wunused-private-field]
    const vector<Function> &outputs;
                            ^
1 error generated.

**Comment by abadams** [comment]
Thanks, pushed what is hopefully a fix.

**Comment by ronlieb** [comment]
thank you, that fixed it

**Pull request merged**